nomos-services/utils/src/overwatch/status/utils.rs

```diff
@@ -0,0 +1,296 @@
+/// Macro call that waits for services to be in `ServiceStatus::Ready` state.
+/// It can wait from 1..n services to be ready.
+///
+/// # Arguments
+///
+/// Requires `RuntimeServiceId` to be defined in the current scope.
+///
+/// * `$overwatch_handle`: A reference to the `OverwatchHandle` that manages the
+///   services.
+/// * `$timeout`: An optional `Duration` that specifies how long to wait for the
+///   services to be ready. If `None`, it will wait indefinitely.
+/// * `$( $service_type:ident ),+`: A list of service types that should be
+///   checked for readiness.
+///
+/// # Returns
+///
+/// - `Ok(())` if all specified services are ready within the timeout.
+/// - `Err(DynError)` if any of the specified services are not ready within the
+///   timeout.
+///
+/// # Example
+///
+/// ```rust,ignore
+/// use std::time::Duration;
+/// use overwatch::overwatch::OverwatchHandle;
+/// use services_utils::overwatch::status::wait_until_services_are_ready;
+///
+/// // The following types would be defined as part of your Overwatch runtime.
+/// struct RuntimeServiceId;
+/// struct ServiceA;
+/// struct ServiceB;
+/// struct ServiceC;
+///
+/// // Mock function to get an OverwatchHandle
+/// fn get_overwatch_handle() -> OverwatchHandle<RuntimeServiceId> {
+///     unimplemented!()
+/// }
+///
+/// fn main() {
+///     let overwatch_handle = get_overwatch_handle();
+///     wait_until_services_are_ready!(
+///        &overwatch_handle,
+///        Some(Duration::from_secs(10)),
+///        ServiceA,
+///        ServiceB,
+///        ServiceC
+///     );
+/// }
+/// ```
+///
+/// A complete example with working service implementations is available in this
+/// module's tests.
+#[macro_export]
+macro_rules! wait_until_services_are_ready {
```

Comment:
Just asking since we've hit a lot of corner cases the `derive_services` macro as well. Does this work with generics, and a mix of generic and non-generic types? Perhaps it makes sense to add a mock generic service in the tests as well to make sure it compiles properly.

Reply:
Sure. Just a test where we mix the two to make sure we don't break it in the future would be sufficient then :)

Reply:
In the existing test case I already included both:
```rust
wait_until_services_are_ready!(
    &self.service_resources_handle.overwatch_handle,
    Some(Duration::from_secs(5)),
    GenericService, // Generic
    ServiceB // Concrete
)?;
```

Reply:
Mhh ok I see. I was more suggesting to have a `Service<Generic>` in there, mostly for syntax reasons. Do you think it's not necessary?